File: 104. CombinationSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class combinationSum {

    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> result = new ArrayList<>();
        backtrack(candidates, target, 0, new ArrayList<>(), result);
        return result;
    }

    private static void backtrack(int[] candidates, int remain, int start, List<Integer> current, List<List<Integer>> result) {
        if (remain == 0) {
            result.add(new ArrayList<>(current));
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            int value = candidates[i];
            if (value > remain) {
                break;
            }

            current.add(value);
            backtrack(candidates, remain - value, i, current, result);
            current.remove(current.size() - 1);
        }
    }

    public static void main(String[] args) {
        System.out.println(combinationSum(new int[] {2, 3, 6, 7}, 7));
        System.out.println(combinationSum(new int[] {2, 3, 5}, 8));
        System.out.println(combinationSum(new int[] {2}, 1));
    }
}

File: 104. CombinationSum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void backtrack(vector<int>& candidates, int remain, int start, vector<int>& current, vector<vector<int>>& result) {
    if (remain == 0) {
        result.push_back(current);
        return;
    }
    for (int i = start; i < (int)candidates.size(); i++) {
        int value = candidates[i];
        if (value > remain) break;
        current.push_back(value);
        backtrack(candidates, remain - value, i, current, result);
        current.pop_back();
    }
}

vector<vector<int>> combinationSum(vector<int> candidates, int target) {
    sort(candidates.begin(), candidates.end());
    vector<vector<int>> result;
    vector<int> current;
    backtrack(candidates, target, 0, current, result);
    return result;
}

int main() {
    auto print = [](const vector<vector<int>>& res) {
        cout << "[";
        for (size_t i = 0; i < res.size(); i++) {
            if (i) cout << ", ";
            cout << "[";
            for (size_t j = 0; j < res[i].size(); j++) {
                if (j) cout << ", ";
                cout << res[i][j];
            }
            cout << "]";
        }
        cout << "]\n";
    };
    print(combinationSum({2, 3, 6, 7}, 7));
    print(combinationSum({2, 3, 5}, 8));
    print(combinationSum({2}, 1));
    return 0;
}
